### PR TITLE
SM-1146: Display total counts of projects, machine accounts, secrets in Secrets Manager

### PR DESCRIPTION
**Pull request by mzieniukbw**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/SM-1146

## 📔 Objective

Display total counts of projects, machine accounts, secrets in Secrets Manager:
- Navigation: Projects, Secrets, Machine accounts
- Organization overview: Projects, Secrets
- Project details: Secrets, People, Machine accounts
- Machine account details: Projects, People, Access Tokens

### Notes
- libs/components/src/tabs/tab-nav-bar/tab-link.component.html
Added explicit end slot, so the total counts can have it's own styles. By default excluding hover underline.
- See server PR https://github.com/bitwarden/server/pull/4200

### Navigation and Organization overview:
image

### Project details
image

### Machine account details
image

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

**Review by willmartian — CHANGES_REQUESTED** [review]
Hey @mzieniukbw! Could you add a story that showcase the new feature? 

@bitwarden/dept-design could this be added to the tab's Figma spec?

**Comment by mzieniukbw** [comment]
> Hey @mzieniukbw! Could you add a story that showcase the new feature?
> 
> @bitwarden/dept-design could this be added to the tab's Figma spec?

@willmartian 

Added with https://github.com/bitwarden/clients/pull/9791/commits/2f743cabbbb1c05cc92ff9e1456811fa9ae91a2e

Screenshot:
image
Note: I noticed a bug that the content of the tab does not display with Light & Dark themes (both) are selected. It was there before i made the change.

File: libs/components/src/tabs/tab-nav-bar/tab-link.component.html [hunk 1]
```diff
@@ -9,6 +9,10 @@
   [attr.aria-disabled]="disabled"
   ariaCurrentWhenActive="page"
   role="link"
+  [class]="'tw-flex' + (haveEndSlot ? ' hover:tw-no-underline' : '')"
 >
-  <ng-content></ng-content>
+  <div [class]="'tw-inline-flex' + (haveEndSlot ? ' hover:tw-underline' : '')">
+    <ng-content></ng-content>
+  </div>
+  <ng-content #endSlot select="[slot=end]"></ng-content>
```
Comment (willmartian) [inline libs/components/src/tabs/tab-nav-bar/tab-link.component.html:17]: I'm don't think extra JS is necessary to achieve these styles:

- We want an underline on the main content when the parent is hovered
- We never want a underline on the end slot content
- We want a left margin on the end slot _only if it has children_ (so it doesn't need to be applied [at the usage site](https://github.com/bitwarden/clients/pull/9791/files#diff-33d3bf7319bddf4d2a7f7da5922767ec5f00224267e70e485cd407d9594ab107R115))

```html
  ...
  role="link"
  class="tw-flex tw-group/tab hover:tw-no-underline"
>
  <div class="group-hover/tab:tw-underline">
    <ng-content></ng-content>
  </div>
  <div class="tw-font-normal tw-ml-2 empty:tw-ml-0">
    <ng-content select="[slot=end]"></ng-content>
  </div>
</a>
```

(Double check this please)

**Review by willmartian — CHANGES_REQUESTED** [review]

**Review by Thomas-Avery — CHANGES_REQUESTED** [review]
Looking good, thank you for all the great unit tests!

A couple of small things to take a look at.

File: bitwarden_license/bit-web/src/app/secrets-manager/projects/project/project.component.ts [hunk 2]
```diff
@@ -62,13 +70,24 @@ export class ProjectComponent implements OnInit, OnDestroy {
     const organization$ = this.route.params.pipe(
       concatMap((params) => this.organizationService.get$(params.organizationId)),
     );
+    const projectCounts$ = combineLatest([
+      this.route.params,
+      this.secretService.secret$.pipe(startWith(null)),
+      this.accessPolicyService.accessPolicy$.pipe(startWith(null)),
+      this.serviceAccountService.serviceAccount$.pipe(startWith(null)),
```
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply bitwarden_license/bit-web/src/app/secrets-manager/projects/project/project.component.ts]: Nice observation! 💯  It does not make much sense to also watch changes on the service accounts, since the only action that can be undertaken that can affect the service account count from the local session and that component, are the access policy changes (service account (un)assigned to/from the project).
Fixed in https://github.com/bitwarden/clients/pull/9791/commits/3c237a8d65bce1fe15f97219360cf05c20f422f3

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets-manager.service.ts
[the remark this answers is not in the document]
Comment (mzieniukbw) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets-manager.service.ts:1]: Nice suggestion. I went step further and combined the count methods and responses from projects and service account services into one service. Let me know if that's ok.
See https://github.com/bitwarden/clients/pull/9791/commits/43a1fdaccf66ec7b8671f4577d5a76b585b25e81

File: libs/components/src/tabs/tab-nav-bar/tab-link.component.html (hunk 1 above)
Comment (mzieniukbw) [reply libs/components/src/tabs/tab-nav-bar/tab-link.component.html]: That works great! Did now know tailwind have such capability, I will have to read more on this!
Fixed in https://github.com/bitwarden/clients/pull/9791/commits/2df6e0294c7448911ba5f38421a960c4ed66a97e

File: bitwarden_license/bit-web/src/app/secrets-manager/secrets-manager.service.ts
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/secrets-manager/secrets-manager.service.ts:1]: Nice! 

I'm a fan of having the service in the client match the controller in the server.

**Review by Thomas-Avery — APPROVED** [review]
LGTM

**Review by willmartian — APPROVED** [review]
Thanks!

**Comment by mzieniukbw** [comment]
@willmartian I had to make the slot conditional, due to a defect (SM-1337) and i have found that the implementation with the `endSlot.childElementCount` causes angular exception. This is because the condition is async, the `div` component renders empty first and then updates, causes the angular change detection, which apparently does not work out of the box - was getting NG0100 complaining about 'tw-hidden' being added.
I could not get this working correctly, since it's a chicken-egg scenario and did not want to go with the programatic way, so i just changed it to `empty:tw-hidden`.
See change in https://github.com/bitwarden/clients/pull/9791/commits/076102992cc88588a0454284523e9dfb9a96fc34

Note that this will actually not hide the slot's content from rendering when `ngIf` is used on the slot, since `empty` selector only works when there is 0 characters (including whitespaces and html comments). Angular community is aware of this [issue](https://github.com/angular/angular/issues/26083) and it actually got fixed with angular [v18 PR](https://github.com/angular/angular/pull/54854).

For reference this is how it looks when `ngIf` is used (does not cause any visible changes as to was before, just empty space):
image

if you have any other idea on how to fix this or you prefer to hide it programatically, i am open to suggestions.

**Comment by mzieniukbw** [comment]
@coltonhurst I have modified the SM import error handling that you have wrote a while ago, since i could not get it working with projects for the Unit Test in https://github.com/bitwarden/clients/pull/9791/commits/bddc739fed9b7e671e4ae067fad85bb4d9c18407
I have also tested the success and edge cases on various scenarios (like missing fields, invalid format of the data) both in projects and secrets and it works fine now.
Appreciate any feedback on this.

File: bitwarden_license/bit-web/src/app/secrets-manager/settings/services/sm-porting-api.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+      expect(subscriptionCount).toEqual(1);
+    });
+
+    it("correct api service send parameters", async () => {
+      const expectedRequest = toRequest([project1, project2], [secret1, secret2]);
+
+      await sut.import("organizationId", importData);
+
+      expect(apiService.send).toBeCalledWith(
```
Comment (Thomas-Avery) [inline bitwarden_license/bit-web/src/app/secrets-manager/settings/services/sm-porting-api.service.spec.ts:76]: Looks like `toBeCalledWith` is marked as deprecated with a note to us `toHaveBeenCalledWith`.

Any reason not to swap over?

**Review by Thomas-Avery — CHANGES_REQUESTED** [review]
Looks good, let's make sure to have QA do regression testing on import. 

Few small things to look at.

**Comment by jtouchstonebw** [comment]
> Hey @mzieniukbw! Could you add a story that showcase the new feature?
> 
> @bitwarden/dept-design could this be added to the tab's Figma spec?

The legacy tailwind figma has been updated with examples. Working on the other figma library atm.

**Review by Thomas-Avery — COMMENTED** [review]
Changes look good.

This will need to have merge conflicts fixed.
I believe those tests were broken on main, the merge should fix that.

**Comment by mzieniukbw** [comment]
I had to run prettier to fix the linting issue which updated loads of files and added a lot of teams as reviewers. I have removed them now from reviewers. I can imagine this would also fail on main ?

**Review by Thomas-Avery — APPROVED** [review]

**Comment by Thomas-Avery** [comment]
> I had to run prettier to fix the linting issue which updated loads of files and added a lot of teams as reviewers. I have removed them now from reviewers. I can imagine this would also fail on main ?

Looks like that is now fixed.

**Comment by jtouchstonebw** [comment]
> > Hey @mzieniukbw! Could you add a story that showcase the new feature?
> > @bitwarden/dept-design could this be added to the tab's Figma spec?
> 
> The legacy tailwind figma has been updated with examples. Working on the other figma library atm.

Following up on this @willmartian - the Figma spec has been updated to show the configuration/variant for including count in tab content.

**Comment by mzieniukbw** [comment]
@willmartian could you approve if everything looks ok ? The changes have been QA tested, waiting for approvals.

**Review by willmartian — APPROVED** [review]
LGTM! (Sorry for the delay.)
